src/modules/commander/Commander.cpp

```diff
@@ -3872,203 +3813,117 @@ bool Commander::preflight_check(bool report)
 				}
 
-			} else if (!_telemetry[i].lost) {
-				/* telemetry was healthy also in last iteration
-				 * we don't have to check a timeout */
-				have_link = true;
 
-				if (!_telemetry[i].high_latency) {
-					have_low_latency_link = true;
-				}
-			}
+				// handle different remote types
+				switch (telemetry.remote_type) {
+				case (telemetry_status_s::MAV_TYPE_GCS):
+					_datalink_last_heartbeat_gcs = telemetry.heartbeat_time;
 
-		} else {
+					if (status.data_link_lost) {
+						if (hrt_elapsed_time(&_datalink_lost) > (_datalink_regain_threshold.get() * 1_s)) {
+							status.data_link_lost = false;
+							status_changed = true;
+						}
+					}
 
-			if (!_telemetry[i].lost) {
-				/* only reset the timestamp to a different time on state change */
-				_telemetry[i].last_dl_loss  = hrt_absolute_time();
+					break;
 
-				mavlink_and_console_log_info(&mavlink_log_pub, "data link #%i lost", i);
-				_telemetry[i].lost = true;
+				case (telemetry_status_s::MAV_TYPE_ONBOARD_CONTROLLER):
+					_datalink_last_heartbeat_onboard_controller = telemetry.heartbeat_time;
+					// TODO: FYI @baumanta
+					break;
+				}
 			}
 		}
 	}
 
-	if (have_link) {
-		/* handle the case where data link was regained */
-		if (status.data_link_lost) {
-			status.data_link_lost = false;
-			*status_changed = true;
-		}
+	// GCS data link loss failsafe
+	if (hrt_elapsed_time(&_datalink_last_heartbeat_gcs) > (_datalink_loss_threshold.get() * 1_s)) {
+		_datalink_lost = hrt_absolute_time();
 
-		if (status.high_latency_data_link_active && have_low_latency_link) {
-			// regained a low latency telemetry link, deactivate the high latency link
-			// to avoid transmitting unnecessary data over that link
-			status.high_latency_data_link_active = false;
-			*status_changed = true;
-			mavlink_log_critical(&mavlink_log_pub, "LOW LAT LINKS REGAINED, DEACTIVATE HIGH LAT LINK");
+		if (!status.data_link_lost) {
+			status.data_link_lost = true;
+			status.data_link_lost_counter++;
+			mavlink_log_critical(&mavlink_log_pub, "DATA LINK LOST");
+			status_changed = true;
 		}
+	}
 
-	} else {
-		if (high_latency_link_exists && !status.high_latency_data_link_active && armed.armed) {
-			// low latency telemetry lost and high latency link existing
-			status.high_latency_data_link_active = true;
-			*status_changed = true;
-
-			// set heartbeat to current time for high latency so that the first message can be transmitted
-			for (int i = 0; i < ORB_MULTI_MAX_INSTANCES; i++) {
-				if (_telemetry[i].high_latency) {
-					_telemetry[i].last_heartbeat = hrt_absolute_time();
-				}
-			}
-
-			if (!status.data_link_lost) {
-				mavlink_log_critical(&mavlink_log_pub, "ALL LOW LAT LINKS LOST, ACTIVATE HIGH LAT LINK");
+	// ONBOARD CONTROLLER data link loss failsafe (hard coded 5 seconds)
+	//  only issue a periodic warning for now
+	if ((_datalink_last_heartbeat_onboard_controller > 0)
+	    && (hrt_elapsed_time(&_datalink_last_heartbeat_onboard_controller) > 5_s) &&
```

Comment:
TODO: review hardcoded 5 seconds.

Reply:
Done in [41b4e20](https://github.com/PX4/Firmware/pull/11550/commits/41b4e20d0f6bd66a4a1f464e82ce40242ddacd11).